src/interpolate.rs

```diff
@@ -0,0 +1,210 @@
+use {Duplex, Frame, Sample};
+
+/// An iterator that converts the rate at which frames are yielded from some given frame
+/// Interpolator into a new type.
+///
+/// Other names for `sample::rate::Converter` might include:
+///
+/// - Sample rate converter
+/// - {Up/Down}sampler
+/// - Sample interpolater
+/// - Sample decimator
+///
+#[derive(Clone)]
+pub struct Converter<T: Iterator, I: Interpolator>
+    where <T as Iterator>::Item: Frame
+{
+    source: T,
+    interpolator: I,
+    interpolation_value: f64,
+    source_to_target_ratio: f64
+}
+
+/// Interpolator that just rounds off any values to the previous value from the source
+pub struct Floor<F>
+{
+    left: F
+}
+
+/// Interpolator that interpolates linearly between the previous value and the next value
+pub struct Linear<F>
+{
+    left: F,
+    right: Option<F>
+}
+
+/// Trait for all things that can interpolate between two values. Implementations should keep track
+/// of the necessary data both before and after the current frame.
+pub trait Interpolator
+{
+    type Frame: Frame;
+
+    /// Given a distance between [0. and 1.) to the following sample, return the interpolated value
+    fn interpolate(&self, x: f64) -> Self::Frame;
```

Comment:
I see...we would need to pass two numbers, then. And the `Duplex<f64>` bound doesn't really materially affect anything, since it's implemented for every sample type. I see your point about efficient interpolation though - we could introduce rounding errors here. But I also think that interface should be driven by the eventual implementation, because otherwise we don't know how ergonomic it will be.